### PR TITLE
Cleanup the deprecated package of  k8s.io/utils/pointer.

### PR DESCRIPTION
**Pull request by liangyuanpeng**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

~~1. Bump k8s.io/utils to 20240102154912-e7106e64919e~~
2. cleanup the deprecated package of pointer, use `k8s.io/utils/ptr` instead.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/resourceinterpreter/customized/declarative/luavm/lua_test.go [hunk 1]
```diff
@@ -641,7 +641,7 @@ func Test_decodeValue(t *testing.T) {
 		{
 			name: "int pointer",
 			args: args{
-				value: pointer.Int(1),
+				value: ptr.To(int)(1),
```
Comment (yanfeng1992) [inline pkg/resourceinterpreter/customized/declarative/luavm/lua_test.go:644]: `value:ptr.To[int](1)` 

@liangyuanpeng

**Comment by XiShanYongYe-Chang** [comment]
Hi @liangyuanpeng the CI is failed.

**Comment by liangyuanpeng** [comment]
```shell
#3 ERROR: failed to do request: Head "https://registry-1.docker.io/v2/library/alpine/manifests/3.19.1": net/http: TLS handshake timeout
------
 > [internal] load metadata for docker.io/library/alpine:3.19.1:
------
Dockerfile:15
--------------------
  13 |     # limitations under the License.
  14 |     
  15 | >>> FROM alpine:3.19.1
  16 |     
  17 |     ARG BINARY
--------------------
ERROR: failed to solve: alpine:3.19.1: failed to do request: Head "https://registry-1.docker.io/v2/library/alpine/manifests/3.19.1": net/http: TLS handshake timeout
make: *** [Makefile:58: image-karmada-operator] Error 1
make: Leaving directory '/home/runner/work/karmada/karmada'
Error: Process completed with exit code 2.
```

CI failed with timeout for pull `alpine:3.19.1`

**Comment by RainbowMango** [comment]
@liangyuanpeng 
Karmada usually keeps Kubernetes dependencies consistent, that's the reason why we use `v0.0.0-20230406110748-d93618cff8a2` (which is consistent with Kubernetes v1.28.5). Given the `k8s.io/utils/pointer` hasn't been deprecated in v1.28.5(deprecated in v1.29), so I think we can postpone this for a while until we bump Kubernetes v1.29. What do you think?

In addition, we are waiting for https://github.com/kubernetes-sigs/custom-metrics-apiserver/issues/167 to bump Kuberentes dependencies v1.29.

**Comment by liangyuanpeng** [comment]
Thank you for your detailed explanation, it's make sense for me.

**Comment by RainbowMango** [comment]
Hi @liangyuanpeng We can get back on this now as the Kubernetes dependencies now is v1.29.4.

> Bump k8s.io/utils to 20240102154912-e7106e64919e

We don't need to bump the version, right? We can cleanup the usage of deprecated packages though.

**Comment by liangyuanpeng** [comment]
I will update it ASAP.

**Comment by RainbowMango** [comment]
Kindly ping @liangyuanpeng 
This is the last blocker of #3835, probably.

> Enable static check by removing [these lines](https://github.com/karmada-io/karmada/blame/e5277b6317ac1a4717f5fac4057caf51a5d248fc/.golangci.yml#L69-L72)

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Pull request merged**